File: java/src/AgentLoop.java
/**
 * Plug-and-play Agentic Loop — Java
 * Multi-provider via ModelRouter (Claude / DeepSeek / OpenAI).
 * All tools are real implementations from ReactLoop.TOOL_REGISTRY — no stubs.
 *
 * export MODEL_PROVIDER=anthropic|deepseek|openai
 * export ANTHROPIC_API_KEY / DEEPSEEK_API_KEY / OPENAI_API_KEY
 * java -cp target/rag-ai-java-1.0.0.jar AgentLoop "your task"
 */

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.*;

import java.util.*;

public class AgentLoop {

    static final ObjectMapper JSON = new ObjectMapper();

    /**
     * Minimal single-shot agentic loop using ModelRouter + ReactLoop.TOOL_REGISTRY.
     * For multi-step ReAct behaviour with Thought/Act/Observe, use ReactLoop.runReact() instead.
     */
    @SuppressWarnings("unchecked")
    public static String agentLoop(
        String userMessage,
        int maxIterations,
        String systemPrompt
    ) throws Exception {
        ModelRouter router = new ModelRouter();
        System.err.printf("[agent_loop] provider=%s model=%s%n", router.provider, router.model);

        List<Map<String, Object>> messages = new ArrayList<>();
        messages.add(Map.of("role", "user", "content", userMessage));

        for (int i = 0; i < maxIterations; i++) {
            System.err.printf("[agent_loop] iteration %d%n", i + 1);

            ModelRouter.UnifiedResponse resp = router.create(
                messages,
                ReactLoop.buildToolSchemas(),
                systemPrompt,
                4096
            );

            messages.add(resp.toAssistantMessage());

            if ("end_turn".equals(resp.stopReason)) {
                for (ModelRouter.ContentBlock block : resp.content) {
                    if (block.type == ModelRouter.BlockType.TEXT && block.text != null && !block.text.isBlank()) {
                        return block.text.trim();
                    }
                }
            }

            if ("tool_use".equals(resp.stopReason)) {
                List<Map<String, Object>> toolResultBlocks = new ArrayList<>();
                for (ModelRouter.ContentBlock block : resp.content) {
                    if (block.type != ModelRouter.BlockType.TOOL_USE) continue;

                    // finish tool — return answer immediately
                    if ("finish".equals(block.name)) {
                        Object answer = block.input != null ? block.input.get("answer") : null;
                        return answer != null ? answer.toString() : "";
                    }

                    String inputStr = JSON.writeValueAsString(block.input);
                    System.err.printf("[agent_loop] ⚡ %s(%s)%n",
                        block.name, inputStr.length() > 120 ? inputStr.substring(0, 120) : inputStr);

                    String result;
                    boolean isError = false;
                    ReactLoop.ToolDef toolDef = ReactLoop.TOOLS.get(block.name);
                    ReactLoop.ToolFn fn = toolDef != null ? toolDef.fn() : null;
                    if (fn == null) {
                        result  = "Tool '" + block.name + "' is not registered";
                        isError = true;
                    } else {
                        try {
                            com.fasterxml.jackson.databind.JsonNode inputNode =
                                block.input != null
                                    ? JSON.valueToTree(block.input)
                                    : JSON.createObjectNode();
                            result = fn.apply(inputNode);
                        } catch (Exception e) {
                            result  = "Tool error (" + block.name + "): " + e.getMessage();
                            isError = true;
                        }
                    }

                    System.err.printf("[agent_loop] 👁 %s%n", result.length() > 120 ? result.substring(0, 120) : result);

                    Map<String, Object> tr = new LinkedHashMap<>();
                    tr.put("type", "tool_result");
                    tr.put("tool_use_id", block.id);
                    tr.put("content", result);
                    if (isError) tr.put("is_error", true);
                    toolResultBlocks.add(tr);
                }
                messages.add(Map.of("role", "user", "content", toolResultBlocks));
                continue;
            }

            System.err.printf("[agent_loop] unexpected stop_reason=%s%n", resp.stopReason);
            break;
        }

        throw new RuntimeException("AgentLoop did not finish within " + maxIterations + " iterations");
    }

    public static void main(String[] args) throws Exception {
        String task = args.length > 0
            ? String.join(" ", args)
            : "Use bash to find the current date and time, then report it.";
        System.err.printf("[agent_loop_demo] task: %s%n", task);
        String answer = agentLoop(task, 10, "You are a helpful assistant. Use provided tools to complete tasks.");
        System.out.println(answer);
    }
}

File: rust/src/agent_loop.rs
// Plug-and-play Agentic Loop — Rust
// Multi-provider via ModelRouter (Claude / DeepSeek / OpenAI).
// All tools are real implementations from ToolRegistry — no stubs.
//
// export MODEL_PROVIDER=anthropic|deepseek|openai
// export ANTHROPIC_API_KEY / DEEPSEEK_API_KEY / OPENAI_API_KEY
// cargo run --bin run_agent -- agent --task "your question"

use std::env;
use anyhow::Result;
use serde_json::json;

use crate::{ModelRouter, ContentBlock, ToolRegistry, make_tool_result_message};

/// Minimal single-shot agentic loop using ModelRouter + ToolRegistry.
/// For multi-step ReAct behaviour with Thought/Act/Observe, use `run_react()` instead.
pub async fn agent_loop(
    user_message: &str,
    max_iterations: usize,
    system: Option<&str>,
) -> Result<String> {
    let router   = ModelRouter::new(None, None)?;
    let registry = ToolRegistry::new();
    let schemas  = registry.schemas();

    let mut messages: Vec<serde_json::Value> = vec![
        json!({ "role": "user", "content": user_message })
    ];

    eprintln!("[agent_loop] provider={} model={}", router.provider, router.model);

    for i in 0..max_iterations {
        eprintln!("[agent_loop] iteration {}", i + 1);

        let resp = router.create(
            &messages,
            Some(&schemas),
            system,
            4096,
        ).await?;

        messages.push(resp.to_assistant_message());

        if resp.stop_reason == "end_turn" {
            for block in &resp.content {
                if let ContentBlock::Text { text } = block {
                    let t = text.trim().to_string();
                    if !t.is_empty() {
                        return Ok(t);
                    }
                }
            }
        }

        if resp.stop_reason == "tool_use" {
            let mut tool_results: Vec<serde_json::Value> = Vec::new();

            for block in &resp.content {
                if let ContentBlock::ToolUse { id, name, input } = block {
                    // finish tool — extract answer and return immediately
                    if name == "finish" {
                        let answer = input.get("answer")
                            .and_then(|v| v.as_str())
                            .unwrap_or("")
                            .to_string();
                        return Ok(answer);
                    }

                    let input_val = serde_json::json!(input);
                    eprintln!("[agent_loop] ⚡ {}({})",
                        name,
                        &input_val.to_string()[..input_val.to_string().len().min(120)]);

                    let (result, is_error) = match registry.call(name, input_val) {
                        Ok(r)  => (r, false),
                        Err(e) => {
                            eprintln!("[agent_loop] ❌ {}: {}", name, e);
                            (format!("Tool error ({}): {}", name, e), true)
                        }
                    };

                    eprintln!("[agent_loop] 👁 {}", &result[..result.len().min(120)]);
                    tool_results.push(make_tool_result_message(id, &result, is_error));
                }
            }

            // Flatten tool results into a single user message with content array
            let content: Vec<serde_json::Value> = tool_results.into_iter().flat_map(|msg| {
                msg["content"].as_array().cloned().unwrap_or_default()
            }).collect();
            messages.push(json!({ "role": "user", "content": content }));
            continue;
        }

        eprintln!("[agent_loop] unexpected stop_reason: {}", resp.stop_reason);
        break;
    }

    anyhow::bail!("AgentLoop did not finish within {} iterations", max_iterations)
}

/// Demo task runner — call from a binary's #[tokio::main] async fn main().
pub async fn agent_loop_demo() -> Result<String> {
    let task = "Use bash to find the current date and time, then report it.";
    eprintln!("[agent_loop_demo] task: {}", task);
    agent_loop(task, 10, Some("You are a helpful assistant. Use provided tools to complete tasks.")).await
}
